### ISSUE TITLE
sam sync doesn't resolve Layer ARN when LayerVersion defined in main stack and Function defined in nested stack

### ISSUE THREAD

**Issue by kdo-arine** [aws/aws-sam-cli#3456]
### Description:

I have a main stack that defines a LayerVersion and a nested stack that defines a Function and Api resources. I can run `sam build` and `sam deploy` and everything works as expected. However, when I try `sam sync --watch --no-dependency-layer` and make code change in my Function or Layer, I get the following error where the Layer ARN isn't being processed properly and its logical ID is used instead. Below, you can see that `WeathermanLayer` is being parsed as a string, instead of being resolved into its actual ARN.

### Steps to reproduce:

To reproduce the bug, see: https://github.com/kdo-arine/weatherman. The repo is self-explanatory.

### Observed result:

```bash
(venv) kdo in ~/tmp/weatherman (main) $ sam sync --watch --no-dependency-layer --debug
2021-11-10 05:20:49,116 | Telemetry endpoint configured to be https://aws-serverless-tools-telemetry.us-west-2.amazonaws.com/metrics
2021-11-10 05:20:49,121 | Telemetry endpoint configured to be https://aws-serverless-tools-telemetry.us-west-2.amazonaws.com/metrics
2021-11-10 05:20:49,123 | Using config file: samconfig.toml, config environment: default
2021-11-10 05:20:49,123 | Expand command line arguments to:
2021-11-10 05:20:49,123 | --template_file=/Users/kdo/tmp/weatherman/templates/weatherman.yml --watch --stack_name=weatherman --s3_prefix=weatherman --capabilities=('CAPABILITY_NAMED_IAM', 'CAPABILITY_AUTO_EXPAND') 

                Managed S3 bucket: aws-sam-cli-managed-default-samclisourcebucket-c7rxm5bvzd2p
                A different default S3 bucket can be set in samconfig.toml
                Or by specifying --s3-bucket explicitly.

                Default capabilities applied: ('CAPABILITY_NAMED_IAM', 'CAPABILITY_AUTO_EXPAND')
To override with customized capabilities, use --capabitilies flag or set it in samconfig.toml
2021-11-10 05:20:49,880 | Using build directory as .aws-sam/build
2021-11-10 05:20:49,880 | Using build directory as .aws-sam/build

This feature is currently in beta. Visit the docs page to learn more about the AWS Beta terms https://aws.amazon.com/service-terms/.

The SAM CLI will use the AWS Lambda, Amazon API Gateway, and AWS StepFunctions APIs to upload your code without 
performing a CloudFormation deployment. This will cause drift in your CloudFormation stack. 
**The sync command should only be used against a development stack**.

Confirm that you are synchronizing a development stack and want to turn on beta features.

Enter Y to proceed with the command, or enter N to cancel:
 [y/N]: y
2021-11-10 05:20:51,498 | No Parameters detected in the template
2021-11-10 05:20:51,527 | 2 stacks found in the template
2021-11-10 05:20:51,537 | Collected default values for parameters: {}
2021-11-10 05:20:51,563 | 2 stacks found in the template
2021-11-10 05:20:51,563 | No Parameters detected in the template
2021-11-10 05:20:51,591 | 2 resources found in the stack 
2021-11-10 05:20:51,591 | Collected default values for parameters: {}
2021-11-10 05:20:51,618 | 2 resources found in the stack Api
2021-11-10 05:20:51,618 | No Parameters detected in the template
2021-11-10 05:20:51,641 | Collected default values for parameters: {}
2021-11-10 05:20:51,666 | Found Serverless function with name='GetWeatherForecastFunction' and CodeUri='../functions/get_weather_forecast'
2021-11-10 05:20:51,667 | --base-dir is not presented, adjusting uri ../functions/get_weather_forecast relative to /Users/kdo/tmp/weatherman/templates/api.yml
2021-11-10 05:20:51,667 | No Parameters detected in the template
2021-11-10 05:20:51,689 | --base-dir is not presented, adjusting uri ../layers/weatherman relative to /Users/kdo/tmp/weatherman/templates/weatherman.yml
2021-11-10 05:20:51,689 | Collected default values for parameters: {}
2021-11-10 05:20:51,718 | Queued infra sync. Wating for in progress code syncs to complete...
2021-11-10 05:20:51,718 | Starting infra sync.
2021-11-10 05:20:51,721 | No Parameters detected in the template
2021-11-10 05:20:51,742 | 2 stacks found in the template
2021-11-10 05:20:51,751 | Collected default values for parameters: {}
2021-11-10 05:20:51,774 | 2 stacks found in the template
2021-11-10 05:20:51,774 | No Parameters detected in the template
2021-11-10 05:20:51,797 | 2 resources found in the stack 
2021-11-10 05:20:51,797 | Collected default values for parameters: {}
2021-11-10 05:20:51,820 | 2 resources found in the stack Api
2021-11-10 05:20:51,820 | No Parameters detected in the template
2021-11-10 05:20:51,843 | Collected default values for parameters: {}
2021-11-10 05:20:51,866 | Found Serverless function with name='GetWeatherForecastFunction' and CodeUri='../functions/get_weather_forecast'
2021-11-10 05:20:51,867 | --base-dir is not presented, adjusting uri ../functions/get_weather_forecast relative to /Users/kdo/tmp/weatherman/templates/api.yml
2021-11-10 05:20:51,867 | No Parameters detected in the template
2021-11-10 05:20:51,889 | --base-dir is not presented, adjusting uri ../layers/weatherman relative to /Users/kdo/tmp/weatherman/templates/weatherman.yml
2021-11-10 05:20:51,889 | Collected default values for parameters: {}
2021-11-10 05:20:51,912 | Instantiating build definitions
2021-11-10 05:20:51,918 | Same function build definition found, adding function (Previous: BuildDefinition(python3.7, /Users/kdo/tmp/weatherman/functions/get_weather_forecast, Zip, , 7726a5e1-2601-430a-af34-479e87f9d21e, {}, {}, x86_64, []), Current: BuildDefinition(python3.7, /Users/kdo/tmp/weatherman/functions/get_weather_forecast, Zip, , 8199283d-724d-41f1-a475-1d024ef3bbf1, {}, {}, x86_64, []), Function: Function(name='GetWeatherForecastFunction', functionname='get_weather_forecast', runtime='python3.7', memory=None, timeout=3, handler='main.lambda_handler', imageuri=None, packagetype='Zip', imageconfig=None, codeuri='/Users/kdo/tmp/weatherman/functions/get_weather_forecast', environment={'Variables': {'MY_API_KEY': 'asdfqwert'}}, rolearn=None, layers=[<samcli.lib.providers.provider.LayerVersion object at 0x1073e8130>], events={'MyApi': {'Type': 'Api', 'Properties': {'Method': 'get', 'Path': '/forecast', 'RestApiId': 'ServerlessRestApi'}}}, metadata=None, inlinecode=None, codesign_config_arn=None, architectures=None, stack_path='Api'))
2021-11-10 05:20:51,918 | Same Layer build definition found, adding layer (Previous: LayerBuildDefinition(WeathermanLayer, /Users/kdo/tmp/weatherman/layers/weatherman, 871a34b5ca5a4888b05d28c4491f7041393552e353565b3768bf1cfd9b554690, 92c0b4f9-7de0-4bdd-a348-6168de069033, makefile, ['python3.7'], x86_64, {}), Current: LayerBuildDefinition(WeathermanLayer, /Users/kdo/tmp/weatherman/layers/weatherman, , d3386812-02b2-4ad5-8ebd-51b9f7d67025, makefile, ['python3.7'], x86_64, {}), Layer: <samcli.lib.providers.provider.LayerVersion object at 0x10755da90>)
2021-11-10 05:20:51,920 | Async execution started
2021-11-10 05:20:51,920 | Invoking function functools.partial(<bound method CachedOrIncrementalBuildStrategyWrapper.build_single_function_definition of <samcli.lib.build.build_strategy.CachedOrIncrementalBuildStrategyWrapper object at 0x1073de3d0>>, <samcli.lib.build.build_graph.FunctionBuildDefinition object at 0x1073de9d0>)
2021-11-10 05:20:51,920 | Running incremental build for runtime python3.7 for build definition 7726a5e1-2601-430a-af34-479e87f9d21e
2021-11-10 05:20:51,921 | Invoking function functools.partial(<bound method CachedOrIncrementalBuildStrategyWrapper.build_single_layer_definition of <samcli.lib.build.build_strategy.CachedOrIncrementalBuildStrategyWrapper object at 0x1073de3d0>>, <samcli.lib.build.build_graph.LayerBuildDefinition object at 0x1073de3a0>)
2021-11-10 05:20:51,921 | Running cached build for runtime makefile for build definition 92c0b4f9-7de0-4bdd-a348-6168de069033
2021-11-10 05:20:51,921 | Waiting for async results
2021-11-10 05:20:51,922 | Building codeuri: /Users/kdo/tmp/weatherman/functions/get_weather_forecast runtime: python3.7 metadata: {} architecture: x86_64 functions: ['Api/GetWeatherForecastFunction']
2021-11-10 05:20:51,922 | Building to following folder /Users/kdo/tmp/weatherman/.aws-sam/build/Api/GetWeatherForecastFunction
2021-11-10 05:20:51,923 | Loading workflow module 'aws_lambda_builders.workflows'
2021-11-10 05:20:51,924 | Valid cache found, copying previously built resources from layer build definition of 92c0b4f9-7de0-4bdd-a348-6168de069033
2021-11-10 05:20:51,924 | Copying artifacts from /Users/kdo/tmp/weatherman/.aws-sam/cache/92c0b4f9-7de0-4bdd-a348-6168de069033 to /Users/kdo/tmp/weatherman/.aws-sam/build/WeathermanLayer
2021-11-10 05:20:51,929 | Registering workflow 'PythonPipBuilder' with capability 'Capability(language='python', dependency_manager='pip', application_framework=None)'
2021-11-10 05:20:51,932 | Registering workflow 'NodejsNpmBuilder' with capability 'Capability(language='nodejs', dependency_manager='npm', application_framework=None)'
2021-11-10 05:20:51,934 | Registering workflow 'RubyBundlerBuilder' with capability 'Capability(language='ruby', dependency_manager='bundler', application_framework=None)'
2021-11-10 05:20:51,936 | Registering workflow 'GoDepBuilder' with capability 'Capability(language='go', dependency_manager='dep', application_framework=None)'
2021-11-10 05:20:51,938 | Registering workflow 'GoModulesBuilder' with capability 'Capability(language='go', dependency_manager='modules', application_framework=None)'
2021-11-10 05:20:51,942 | Registering workflow 'JavaGradleWorkflow' with capability 'Capability(language='java', dependency_manager='gradle', application_framework=None)'
2021-11-10 05:20:51,945 | Registering workflow 'JavaMavenWorkflow' with capability 'Capability(language='java', dependency_manager='maven', application_framework=None)'
2021-11-10 05:20:51,947 | Registering workflow 'DotnetCliPackageBuilder' with capability 'Capability(language='dotnet', dependency_manager='cli-package', application_framework=None)'
2021-11-10 05:20:51,950 | Registering workflow 'CustomMakeBuilder' with capability 'Capability(language='provided', dependency_manager=None, application_framework=None)'
2021-11-10 05:20:51,950 | Found workflow 'PythonPipBuilder' to support capabilities 'Capability(language='python', dependency_manager='pip', application_framework=None)'
2021-11-10 05:20:51,950 | requirements.txt file not found. Continuing the build without dependencies.
2021-11-10 05:20:51,981 | Running workflow 'PythonPipBuilder'
2021-11-10 05:20:51,982 | Running PythonPipBuilder:CopySource
2021-11-10 05:20:51,987 | PythonPipBuilder:CopySource succeeded
2021-11-10 05:20:51,988 | Async execution completed

Build Succeeded

Built Artifacts  : .aws-sam/build
Built Template   : .aws-sam/build/template.yaml

Commands you can use next
=========================
[*] Invoke Function: sam local invoke
[*] Deploy: sam deploy --guided
        

Successfully packaged artifacts and wrote output template to file /var/folders/t4/95jlxd9n1d91lpc2q4vkqvv80000gn/T/tmplv24rdt0.
Execute the following command to deploy the packaged template
sam deploy --template-file /var/folders/t4/95jlxd9n1d91lpc2q4vkqvv80000gn/T/tmplv24rdt0 --stack-name <YOUR STACK NAME>


        Deploying with following values
        ===============================
        Stack name                   : weatherman
        Region                       : us-east-1
        Disable rollback             : False
        Deployment s3 bucket         : aws-sam-cli-managed-default-samclisourcebucket-c7rxm5bvzd2p
        Capabilities                 : ["CAPABILITY_NAMED_IAM", "CAPABILITY_AUTO_EXPAND"]
        Parameter overrides          : {}
        Signing Profiles             : null

Initiating deployment
=====================
2021-11-10 05:20:53,038 | No Parameters detected in the template
2021-11-10 05:20:53,062 | 2 stacks found in the template
2021-11-10 05:20:53,063 | No Parameters detected in the template
2021-11-10 05:20:53,087 | 2 resources found in the stack 
2021-11-10 05:20:53,087 | No Parameters detected in the template

2021-11-10 05:20:54 - Waiting for stack create/update to complete

CloudFormation events from stack operations
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------
ResourceStatus                             ResourceType                               LogicalResourceId                          ResourceStatusReason                     
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------
UPDATE_IN_PROGRESS                         AWS::CloudFormation::Stack                 weatherman                                 Transformation succeeded                 
UPDATE_IN_PROGRESS                         AWS::CloudFormation::Stack                 Api                                        -                                        
UPDATE_COMPLETE                            AWS::CloudFormation::Stack                 Api                                        -                                        
UPDATE_COMPLETE_CLEANUP_IN_PROGRESS        AWS::CloudFormation::Stack                 weatherman                                 -                                        
UPDATE_COMPLETE                            AWS::CloudFormation::Stack                 weatherman                                 -                                        
UPDATE_COMPLETE                            AWS::CloudFormation::Stack                 Api                                        -                                        
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------
2021-11-10 05:21:56,385 | 
Stack update succeeded. Sync infra completed.

2021-11-10 05:21:56,385 | {'StackId': 'arn:aws:cloudformation:us-east-1:929051535035:stack/weatherman/4b077360-4227-11ec-9ff3-0a8bbda23b13', 'ResponseMetadata': {'RequestId': '18200a56-13da-48bd-b522-0e05fae6bd76', 'HTTPStatusCode': 200, 'HTTPHeaders': {'x-amzn-requestid': '18200a56-13da-48bd-b522-0e05fae6bd76', 'content-type': 'text/xml', 'content-length': '380', 'date': 'Wed, 10 Nov 2021 13:20:54 GMT'}, 'RetryAttempts': 0}}
2021-11-10 05:21:56,390 | No Parameters detected in the template
2021-11-10 05:21:56,416 | 2 stacks found in the template
2021-11-10 05:21:56,425 | Collected default values for parameters: {}
2021-11-10 05:21:56,449 | 2 stacks found in the template
2021-11-10 05:21:56,449 | Loading physical ID mapping
2021-11-10 05:21:56,450 | Fetching stack (weatherman) resources
2021-11-10 05:21:57,007 | No Parameters detected in the template
2021-11-10 05:21:57,030 | Collected default values for parameters: {}
2021-11-10 05:21:57,055 | No Parameters detected in the template
2021-11-10 05:21:57,076 | No Parameters detected in the template
2021-11-10 05:21:57,101 | No Parameters detected in the template
2021-11-10 05:21:57,124 | No Parameters detected in the template
2021-11-10 05:21:57,145 | No Parameters detected in the template
2021-11-10 05:21:57,167 | --base-dir is not presented, adjusting uri ../layers/weatherman relative to /Users/kdo/tmp/weatherman/templates/weatherman.yml
2021-11-10 05:21:57,167 | Collected default values for parameters: {}
2021-11-10 05:21:57,192 | No Parameters detected in the template
2021-11-10 05:21:57,213 | No Parameters detected in the template
2021-11-10 05:21:57,235 | No Parameters detected in the template
2021-11-10 05:21:57,256 | Collected default values for parameters: {}
2021-11-10 05:21:57,278 | Collected default values for parameters: {}
2021-11-10 05:21:57,345 | Collected default values for parameters: {}
2021-11-10 05:21:57,373 | Collected default values for parameters: {}
2021-11-10 05:21:57,441 | No Parameters detected in the template
2021-11-10 05:21:57,464 | 2 resources found in the stack 
2021-11-10 05:21:57,464 | Collected default values for parameters: {}
2021-11-10 05:21:57,488 | 2 resources found in the stack Api
2021-11-10 05:21:57,488 | No Parameters detected in the template
2021-11-10 05:21:57,510 | Collected default values for parameters: {}
2021-11-10 05:21:57,534 | Found Serverless function with name='GetWeatherForecastFunction' and CodeUri='../functions/get_weather_forecast'
2021-11-10 05:21:57,534 | --base-dir is not presented, adjusting uri ../functions/get_weather_forecast relative to /Users/kdo/tmp/weatherman/templates/api.yml
2021-11-10 05:21:57,538 | Collected default values for parameters: {}
2021-11-10 05:21:57,561 | Collected default values for parameters: {}
2021-11-10 05:21:57,583 | Collected default values for parameters: {}
2021-11-10 05:21:57,605 | Collected default values for parameters: {}
2021-11-10 05:21:57,628 | CodeTrigger not created as CodeUri or DefinitionUri is missing for Api/ServerlessRestApi.
2021-11-10 05:21:57,628 | Infra sync completed.
2021-11-10 05:22:37,132 | Collected default values for parameters: {}
2021-11-10 05:22:37,159 | Collected default values for parameters: {}
Exception in thread Thread-1:
Traceback (most recent call last):
  File "/usr/local/Cellar/aws-sam-cli/1.35.0/libexec/lib/python3.8/site-packages/samcli/lib/providers/provider.py", line 249, in _compute_layer_name
    _, layer_name, layer_version = arn.rsplit(":", 2)
ValueError: not enough values to unpack (expected 3, got 1)

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/local/Cellar/python@3.8/3.8.12_1/Frameworks/Python.framework/Versions/3.8/lib/python3.8/threading.py", line 932, in _bootstrap_inner
    self.run()
  File "/usr/local/Cellar/aws-sam-cli/1.35.0/libexec/lib/python3.8/site-packages/watchdog/observers/api.py", line 199, in run
    self.dispatch_events(self.event_queue, self.timeout)
  File "/usr/local/Cellar/aws-sam-cli/1.35.0/libexec/lib/python3.8/site-packages/watchdog/observers/api.py", line 372, in dispatch_events
    handler.dispatch(event)
  File "/usr/local/Cellar/aws-sam-cli/1.35.0/libexec/lib/python3.8/site-packages/watchdog/events.py", line 403, in dispatch
    super().dispatch(event)
  File "/usr/local/Cellar/aws-sam-cli/1.35.0/libexec/lib/python3.8/site-packages/watchdog/events.py", line 271, in dispatch
    self.on_any_event(event)
  File "/usr/local/Cellar/aws-sam-cli/1.35.0/libexec/lib/python3.8/site-packages/samcli/lib/sync/watch_manager.py", line 215, in on_code_change
    self._sync_flow_executor.add_delayed_sync_flow(sync_flow, dedup=True, wait_time=DEFAULT_WAIT_TIME)
  File "/usr/local/Cellar/aws-sam-cli/1.35.0/libexec/lib/python3.8/site-packages/samcli/lib/sync/continuous_sync_flow_executor.py", line 116, in add_delayed_sync_flow
    self._add_sync_flow_task(DelayedSyncFlowTask(sync_flow, dedup, time.time(), wait_time))
  File "/usr/local/Cellar/aws-sam-cli/1.35.0/libexec/lib/python3.8/site-packages/samcli/lib/sync/continuous_sync_flow_executor.py", line 101, in _add_sync_flow_task
    super()._add_sync_flow_task(task)
  File "/usr/local/Cellar/aws-sam-cli/1.35.0/libexec/lib/python3.8/site-packages/samcli/lib/sync/sync_flow_executor.py", line 145, in _add_sync_flow_task
    task.sync_flow.set_locks_with_distributor(self._lock_distributor)
  File "/usr/local/Cellar/aws-sam-cli/1.35.0/libexec/lib/python3.8/site-packages/samcli/lib/sync/sync_flow.py", line 152, in set_locks_with_distributor
    self.set_locks_with_dict(distributor.get_locks(self.get_lock_keys()))
  File "/usr/local/Cellar/aws-sam-cli/1.35.0/libexec/lib/python3.8/site-packages/samcli/lib/sync/sync_flow.py", line 139, in get_lock_keys
    for resource_api_calls in self._get_resource_api_calls():
  File "/usr/local/Cellar/aws-sam-cli/1.35.0/libexec/lib/python3.8/site-packages/samcli/lib/sync/flows/zip_function_sync_flow.py", line 150, in _get_resource_api_calls
    resource_calls.append(ResourceAPICall(layer.full_path, ["Build"]))
  File "/usr/local/Cellar/aws-sam-cli/1.35.0/libexec/lib/python3.8/site-packages/samcli/lib/providers/provider.py", line 324, in full_path
    return get_full_path(self.stack_path, self.name)
  File "/usr/local/Cellar/aws-sam-cli/1.35.0/libexec/lib/python3.8/site-packages/samcli/lib/providers/provider.py", line 282, in name
    self._name = LayerVersion._compute_layer_name(self.is_defined_within_template, self.arn)
  File "/usr/local/Cellar/aws-sam-cli/1.35.0/libexec/lib/python3.8/site-packages/samcli/lib/providers/provider.py", line 251, in _compute_layer_name
    raise InvalidLayerVersionArn(arn + " is an Invalid Layer Arn.") from ex
samcli.commands.local.cli_common.user_exceptions.InvalidLayerVersionArn: WeathermanLayer is an Invalid Layer Arn.
```

### Expected result:

I expect `sam sync` to work as advertised. It *does* work when the LayerVersion (WeathermanLayer) and Function (GetWeatherForecastFunction) resources are defined in the main stack.

### Additional environment details (Ex: Windows, Mac, Amazon Linux etc)

1. OS: macOS Big Sur v11.6
2. `sam --version`: 1.35.0
3. AWS region: us-east-1

`Add --debug flag to command you are running`

**Comment by CoshUS** [comment]
Hey @kdo-arine,
Thanks for the feedback! `sam sync` currently does not support nested stacks, but we are looking into adding support for this use case.

**Comment by kdo-arine** [comment]
Is there room or need to help with this? I would like to contribute if you could give me some pointers. Thanks.

**Comment by ssenchenko** [comment]
We have a [roadmap](https://github.com/aws/aws-sam-cli/wiki/SAM-CLI-Roadmap) here, but we also try to take in popular feature requests based on 👍🏽 from the community. We will be adding this to our intake for prioritization.

**Comment by mndeveci** [comment]
@kdo-arine `sam sync` nested stack support is partially out with current version of SAM CLI. Can you give it a try and give us feedback if you encounter any issues with it?

Thanks!

**Comment by qingchm** [comment]
I believe that this is addressed in the new 1.53.0 release. Please have a try and download our latest version of SAMCLI. If you still have any questions or issues feel free to re-open the issue!
